entities/cyclemanager/cyclecallbackctrl.go

```diff
@@ -0,0 +1,208 @@
+	ctrls         map[string]CycleCallbackCtrl
+}
+
+func NewCycleCombinedCallbackCtrl(ctrls map[string]CycleCallbackCtrl, routinesLimit int) CycleCallbackCtrl {
+	if routinesLimit <= 0 {
+		routinesLimit = _NUMCPU
+	}
+
+	return &cycleCombinedCallbackCtrl{ctrls: ctrls, routinesLimit: routinesLimit}
+}
+
+func (c *cycleCombinedCallbackCtrl) IsActive() bool {
+	for _, ctrl := range c.ctrls {
+		if !ctrl.IsActive() {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *cycleCombinedCallbackCtrl) Activate() error {
+	return c.combineErrors(c.activate()...)
+}
+
+func (c *cycleCombinedCallbackCtrl) activate() []error {
+	eg := &errgroup.Group{}
+	eg.SetLimit(c.routinesLimit)
+	lock := new(sync.Mutex)
+
+	errs := make([]error, 0, len(c.ctrls))
+	for name, ctrl := range c.ctrls {
+		name, ctrl := name, ctrl
+		eg.Go(func() error {
+			if err := ctrl.Activate(); err != nil {
+				err = fmt.Errorf("activating '%s' failed: %w", name, err)
+				c.locked(lock, func() { errs = append(errs, err) })
+				return err
+			}
+			return nil
+		})
+	}
+
+	eg.Wait()
+	return errs
+}
+
+func (c *cycleCombinedCallbackCtrl) Deactivate(ctx context.Context) error {
+	errs, deactivated := c.deactivate(ctx)
+	if len(errs) == 0 {
+		return nil
+	}
+
+	// try activating back deactivated
+	eg := &errgroup.Group{}
+	eg.SetLimit(c.routinesLimit)
+	for _, name := range deactivated {
+		name := name
+		eg.Go(func() error {
+			return c.ctrls[name].Activate()
+		})
+	}
+
+	eg.Wait()
+	return c.combineErrors(errs...)
+}
+
+func (c *cycleCombinedCallbackCtrl) deactivate(ctx context.Context) ([]error, []string) {
+	eg := &errgroup.Group{}
+	eg.SetLimit(c.routinesLimit)
+	lock := new(sync.Mutex)
+
+	errs := make([]error, 0, len(c.ctrls))
+	deactivated := make([]string, 0, len(c.ctrls))
+	for name, ctrl := range c.ctrls {
+		name, ctrl := name, ctrl
+		eg.Go(func() error {
+			if err := ctrl.Deactivate(ctx); err != nil {
+				err = fmt.Errorf("deactivating '%s' failed: %w", name, err)
+				c.locked(lock, func() { errs = append(errs, err) })
+				return err
+			}
+			c.locked(lock, func() { deactivated = append(deactivated, name) })
+			return nil
+		})
+	}
+
+	eg.Wait()
+	return errs, deactivated
+}
+
+func (c *cycleCombinedCallbackCtrl) Unregister(ctx context.Context) error {
+	return c.combineErrors(c.unregister(ctx)...)
+}
+
+func (c *cycleCombinedCallbackCtrl) unregister(ctx context.Context) []error {
+	eg := &errgroup.Group{}
+	eg.SetLimit(c.routinesLimit)
+	lock := new(sync.Mutex)
+
+	errs := make([]error, 0, len(c.ctrls))
+	for name, ctrl := range c.ctrls {
+		name, ctrl := name, ctrl
+		eg.Go(func() error {
+			if err := ctrl.Unregister(ctx); err != nil {
+				err = fmt.Errorf("unregistering '%s' failed: %w", name, err)
+				c.locked(lock, func() { errs = append(errs, err) })
+				return err
+			}
+			return nil
+		})
+	}
+
+	eg.Wait()
+	return errs
+}
+
+func (c *cycleCombinedCallbackCtrl) locked(lock *sync.Mutex, mutate func()) {
+	lock.Lock()
+	mutate()
+	lock.Unlock()
```

Comment:
would be better to use a `defer` here:
``
lock.Lock()
defer lock.Unlock()
mutate()
```
bc now when there would occur a panic inside `mutate()` would we release a lock in this situation?

Reply:
changed